### PR TITLE
ramips: add support for ZTE MF283+

### PR DESCRIPTION
**Pull request by Leo-PL**
ZTE MF283+ is a dual-antenna LTE category 4 router, based on Ralink
RT3352 SoC, and built-in ZTE P685M PCIe MiniCard LTE modem.

Hardware highlighs:
- CPU: MIPS24KEc at 400MHz,
- RAM: 64MB DDR2,
- Flash: 16MB SPI,
- Ethernet: 4 10/100M port switch with VLAN support,
- Wireless: Dual-stream 802.11n (RT2860), with two internal antennas,
- WWAN: Built-in ZTE P685M modem, with two internal antennas and two
  switching SMA connectors for external antennas,
- FXS: Single ATA, with two connectors marked PHONE1 and PHONE2,
  internally wired in parallel by 0-Ohm resistors, handled entirely by
  internal WWAN modem.
- USB: internal miniPCIe slot for modem,
  unpopulated USB A connector on PCB.
- SIM slot for the WWAN modem.
- UART connector for the console (unpopulated) at 3.3V,
  pinout: 1: VCC, 2: TXD, 3: RXD, 4: GND,
  settings: 57600-8-N-1.
- LEDs: Power (fixed), WLAN, WWAN (RGB),
  phone (bicolor, controlled by modem), Signal,
  4 link/act LEDs for LAN1-4.
- Buttons: WPS, reset.

Installation:
As the modem is, for most of the time, provided by carriers, there is no
possibility to flash through web interface, only built-in FOTA update
and TFTP recovery are supported.

There are two installation methods:
(1) Using serial console and initramfs-kernel - recommended, as it
allows you to back up original firmware, or
(2) Using TFTP recovery - does not require disassembly.

(1) Using serial console:
To install OpenWrt, one needs to disassemble the
router and flash it via TFTP by using serial console:
- Locate unpopulated 4-pin header on the top of the board, near buttons.
- Connect UART adapter to the connector. Use 3.3V voltage level only,
  omit VCC connection. Pin 1 (VCC) is marked by square pad.
- Put your initramfs-kernel image in TFTP server directory.
- Power-up the device.
- Press "1" to load initramfs image to RAM.
- Enter IP address chosen for the device (defaults to 192.168.0.1).
- Enter TFTP server IP address (defaults to 192.168.0.22).
- Enter image filename as put inside TFTP server - something short,
  like firmware.bin is recommended.
- Hit enter to load the image. U-boot will store above values in
  persistent environment for next installation.
- If you ever might want to return to vendor firmware,
  BACK UP CONTENTS OF YOUR FLASH NOW.
  For this router, commonly used by mobile networks,
  plain vendor images are not officially available.
  To do so, copy contents of each /dev/mtd[0-3], "firmware" - mtd3 being the
  most important, and copy them over network to your PC. But in case
  anything goes wrong, PLEASE do back up ALL OF THEM.
- From under OpenWrt just booted, load the sysupgrade image to tmpfs,
  and execute sysupgrade.

(2) Using TFTP recovery
- Set your host IP to 192.168.0.22 - for example using:
sudo ip addr add 192.168.0.22/24 dev <interface>
- Set up a TFTP server on your machine
- Put the sysupgrade image in TFTP server root named as 'root_uImage'
  (no quotes), for example using tftpd:
  cp openwrt-ramips-rt305x-zte_mf283plus-squashfs-sysupgrade.bin /srv/tftp/root_uImage
- Power on the router holding BOTH Reset and WPS buttons held for around
  5 seconds, until after WWAN and Signal LEDs blink.
- Wait for OpenWrt to start booting up, this should take around a
  minute.

Return to original firmware:
Here, again there are two possibilities are possible, just like for
installation:
(1) Using initramfs-kernel image and serial console
(2) Using TFTP recovery

(1) Using initramfs-kernel image and serial console
- Boot OpenWrt initramfs-kernel image via TFTP the same as for
  installation.
- Copy over the backed up "firmware.bin" image of "mtd3" to /tmp/
- Use "mtd write /tmp/firmware.bin /dev/mtd3", where firmware.bin is
  your backup taken before OpenWrt installation, and /dev/mtd3 is the
  "firmware" partition.

(2) Using TFTP recovery
- Follow the same steps as for installation, but replacing 'root_uImage'
  with firmware backup you took during installation, or by vendor
  firmware obtained elsewhere.

A few quirks of the device, noted from my instance:
- Wired and wireless MAC addresses written in flash are the same,
  despite being in separate locations.
- Power LED is hardwired to 3.3V, so there is no status LED per se, and
  WLAN LED is controlled by WLAN driver, so I had to hijack 3G/4G LED
  for status - original firmware also does this in bootup.
- FXS subsystem and its LED is controlled by the
  modem, so it work independently of OpenWrt.
  Tested to work even before OpenWrt booted.
  I managed to open up modem's shell via ADB,
  and found from its kernel logs, that FXS and its LED is indeed controlled
  by modem.
- While finding LEDs, I had no GPL source drop from ZTE, so I had to probe for
  each and every one of them manually, so this might not be complete -
  it looks like bicolor LED is used for FXS, possibly to support
  dual-ported variant in other device sharing the PCB.
- Flash performance is very low, despite enabling 50MHz clock and fast
  read command, due to using 4k sectors throughout the target. I decided
  to keep it at the moment, to avoid breaking existing devices - I
  identified one potentially affected, should this be limited to under
  4MB of Flash. The difference between sysupgrade durations is whopping
  3min vs 8min, so this is worth pursuing.

In vendor firmware, WWAN LED behaviour is as follows, citing the manual:
- red - no registration,
- green - 3G,
- blue - 4G.
Blinking indicates activity, so netdev trigger mapped from wwan0 to blue:wwan
looks reasonable at the moment, for full replacement, a script similar to
"rssileds" would need to be developed.

Behaviour of "Signal LED" in vendor firmware is as follows:
- Off - no signal,
- Blinking - poor coverage
- Solid - good coverage.

A few more details on the built-in LTE modem:
Modem is not fully supported upstream in Linux - only two CDC ports
(DIAG and one for QMI) probe. I sent patches upstream to add required device
IDs for full support.
The mapping of USB functions is as follows:
- CDC (QCDM) - dedicated to comunicating with proprietary Qualcomm tools.
- CDC (PCUI) - not supported by upstream 'option' driver yet. Patch
  submitted upstream.
- CDC (Modem) - Exactly the same as above
- QMI - A patch is sent upstream to add device ID, with that in place,
  uqmi did connect successfully, once I selected correct PDP context
  type for my SIM (IPv4-only, not default IPv4v6).
- ADB - self-explanatory, one can access the ADB shell with a device ID
  added to 51-android.rules like so:

SUBSYSTEM!="usb", GOTO="android_usb_rules_end"
LABEL="android_usb_rules_begin"
SUBSYSTEM=="usb", ATTR{idVendor}=="19d2", ATTR{idProduct}=="1275", ENV{adb_user}="yes"
ENV{adb_user}=="yes", MODE="0660", GROUP="plugdev", TAG+="uaccess"
LABEL="android_usb_rules_end"

While not really needed in OpenWrt, it might come useful if one decides to
move the modem to their PC to hack it further, insides seem to be pretty
interesting. ADB also works well from within OpenWrt without that. O
course it isn't needed for normal operation, so I left it out of
DEVICE_PACKAGES.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

### CODE REVIEW

**Comment by Leo-PL** [comment]
@bmork I posted two patches needed for full modem support in this device to linux-usb mailing list earlier today, they are included inside the PR as well. 

If anyone has ZTE MF283V available and could test this too, I'd be very grateful for that.

File: target/linux/ramips/dts/rt3352_zte_mf283plus.dts
Comment (adschm) [inline target/linux/ramips/dts/rt3352_zte_mf283plus.dts:1]: Please add an SPDX license identifier.

File: target/linux/ramips/dts/rt3352_zte_mf283plus.dts
Comment (adschm) [inline target/linux/ramips/dts/rt3352_zte_mf283plus.dts:22]: Please be consistent here, either wwan or 3g4g ...

File: target/linux/ramips/rt305x/base-files/etc/board.d/02_network
Comment (adschm) [inline target/linux/ramips/rt305x/base-files/etc/board.d/02_network:169]: Is the right side of the numbering really necessary?

I'd expect this to give the same result in LuCI:

`"0:lan" "1:lan" "2:lan" "3:lan" "6@eth0"`

**Review by adschm — CHANGES_REQUESTED** [review]

File: target/linux/ramips/rt305x/base-files/etc/board.d/02_network
Comment (Leo-PL) [reply target/linux/ramips/rt305x/base-files/etc/board.d/02_network]: Indeed it does, fixed.

File: target/linux/ramips/dts/rt3352_zte_mf283plus.dts
Comment (Leo-PL) [reply target/linux/ramips/dts/rt3352_zte_mf283plus.dts]: Good catch, fixed.

File: target/linux/ramips/dts/rt3352_zte_mf283plus.dts
Comment (Leo-PL) [reply target/linux/ramips/dts/rt3352_zte_mf283plus.dts:2]: Added.

**Comment by Leo-PL** [comment]
@adschm fixed.

**Comment by mastier** [comment]
awesome piece @Leo-PL !!! Full support, very nice!

**Comment by Leo-PL** [comment]
Changes since previous patchset:
- dropped led-running from DTS in favor of "netdev" trigger on wwan0, to mimic original firmware behaviour a bit better. To mimic it fully, a custom script interacting with `uqmi` would need to be developed from scratch, IIRC.
- Included v2 of patchset for modem included (https://lore.kernel.org/linux-usb/c0a94dc9-cb3d-090c-6e8d-d120de91d5fd@gmail.com/T/#t) and refreshed against kernel 5.4 currently on target. This avoids potential conflict between `option` and `qmi_wwan` drivers on the QMI port.
- Updated documentation in commit message - ADB also works on the device, though not needed outside of hacking.

File: target/linux/ramips/dts/rt3352_zte_mf283plus.dts
Comment (adschm) [inline target/linux/ramips/dts/rt3352_zte_mf283plus.dts:15]: This only works if you assign the address with mtd-mac-address. Use label_mac in 02_network instead.

Comment (Leo-PL) [reply target/linux/ramips/dts/rt3352_zte_mf283plus.dts:15]: Or bind it to Ethernet device device instead - as stated in commit message, both addresses in 'factory' partition were identical, despite being in different places. I determined which one is which by patching this partition.

While we are at that, maybe we should artificially shift one of the addresses? OTOH, as there is no "wan" port per se on this device, other than `wwan0`, and additional "WAN" label on LAN1 port used in non-default configuration in vendor image, it shouldn't hurt to leave them identical, as both are typically on same interface.

What do you think?

Comment (adschm) [reply target/linux/ramips/dts/rt3352_zte_mf283plus.dts:15]: > Or bind it to Ethernet device device instead - as stated in commit message, both addresses in 'factory' partition were identical, despite being in different places. I determined which one is which by patching this partition.

If that's the case, then yes, label-mac-device is preferred where possible.

> While we are at that, maybe we should artificially shift one of the addresses?

It's policy to mirror the vendor assignment as far as possible and reasonable. Changes may always be done by the users later.

Comment (Leo-PL) [reply target/linux/ramips/dts/rt3352_zte_mf283plus.dts:15]: Ok, so let's use Ethernet as label device then. I'll post update shortly, of course I need to rebuild and test, and sysupgrade takes a looooong time on this device at the moment.

Comment (adschm) [reply target/linux/ramips/dts/rt3352_zte_mf283plus.dts:15]: If you just change the label-mac-device, I could live without a test.

Comment (Leo-PL) [reply target/linux/ramips/dts/rt3352_zte_mf283plus.dts:15]: I have a fast machine dedicated for building, so it's still a matter of minutes ;-)

Comment (Leo-PL) [reply target/linux/ramips/dts/rt3352_zte_mf283plus.dts]:
```
root@OpenWrt:~# source /lib/functions/system.sh 
root@OpenWrt:~# get_mac_label
8c:79:67:70:25:c2
root@OpenWrt:~#
```
It works, so resolving.

**Comment by Leo-PL** [comment]
Updated, changed label MAC device to Ethernet, since the address is exactly the same, now `get_mac_label` works.

**Comment by Leo-PL** [comment]
Patch `usb: serial: option: add full support for ZTE P685M` is queued for 5.12-rc1, I'll resubmit patch for qmi_wwan once this one lands in net-netx.

**Comment by Leo-PL** [comment]
Patch `[PATCH v3] usb: serial: option: update interface mapping for ZTE P685M` was queued for -next: https://www.spinics.net/lists/netdev/msg720515.html 
Now waiting for it to hit 'net' to send remaining one for qmi_wwan.

**Comment by Leo-PL** [comment]
@adschm kind reminder.

**Comment by Leo-PL** [comment]
@adschm Do I guess right, that this PR waiting on both patches supporting LTE modem to be merged upstream?

**Comment by KazuDante89** [comment]
@Leo-PL  i hope that im wrong but i wanted to see if my MF283V qualified since its been giving me issues in the last few weeks , i reopen it and i dont see a 4-pin header on the board . could not also find a picture of the mf283v board to verify that my board is the same . in my current situation i really dont want to buy a different modem all over since this one was working fine since 2018 up until recently and was hoping flashing a different firmware would help bring it back to life with some new features from openwrt . also please show me which UART adapter you used on MF283+. will def buy one to test if somehow we can figure out how to flash it on MF283V .

**Comment by Leo-PL** [comment]
> 
> 
> @Leo-PL i hope that im wrong but i wanted to see if my MF283V qualified since its been giving me issues in the last few weeks , i reopen it and i dont see a 4-pin header on the board . could not also find a picture of the mf283v board to verify that my board is the same . in my current situation i really dont want to buy a different modem all over since this one was working fine since 2018 up until recently and was hoping flashing a different firmware would help bring it back to life with some new features from openwrt . also please show me which UART adapter you used on MF283+. will def buy one to test if somehow we can figure out how to flash it on MF283V .

@KazuDante89 Please post photos of your board. I'll post mine later today, so we can cross-check them. The console header is unpopulated, just look for 4 through-hole pads on the very edge of the board, near the two top buttons. Pin 1 is marked by square pad.

**Comment by KazuDante89** [comment]
@Leo-PL does not look to be present on MF283V board , only see 3 possible places ( I am a noob in this department ) but here's the front and back of the board. Sry network so bad I'm having a hard time to upload pictures .

**Comment by Leo-PL** [comment]
Wow. I'm quite surprised to see this. There's more to that than meets the eye. This is entirely different hardware, at first glance I see that Wi-Fi is handled by RTL8192E. It looks like UART is right over USB socket on your board, but you will have to probe this yourself. And I thought that it only differs by a few unpopulated components...

**Comment by KazuDante89** [comment]
@Leo-PL any suggestion on what i need to have to probe this router , because i really would like to help your project because i think openwrt might solve the issues im having with this router , plus might openwrt might add some features like band/freq locking and ttl , because out of nowhere lte is un-usuable and 3g as well and where the router was for the last 3 years it always had 5bars , around the -92 to -95dbm with a decent 14-18mbps down and 25-32(sometimes) up. Now i cant even get a google.com loaded...refreshing over and over until i get luck and thats on 3G . Some site stated that the sma antenna ports are only for 3G so i dont want  make a bad purchase so now im looking into a yeacomm yf-p21 modem and hope it was the mf283v that was the issue .

**Comment by KazuDante89** [comment]
> Wow. I'm quite surprised to see this. There's more to that than meets the eye. This is entirely different hardware, at first glance I see that Wi-Fi is handled by RTL8192E, and router itself is based on Qualcomm Atheros chipset. I don't see on which one though, because of photo resolution.
> 
> It looks like UART is right over USB socket on your board, but you will have to probe this yourself. And I thought that it only differs by a few unpopulated components...
> I guess I'll edit the commit message to remove this reference, perhaps there is yet another router which has missing components from my board populated.

Ill try to get a more close-up shot of the board so that you may see wich chipsets are present on the board . but does that also mean that your openwrt project can't support the mf283V due to so much differences from the MF283+ ?

**Comment by Leo-PL** [comment]
> 
> 
> > Wow. I'm quite surprised to see this. There's more to that than meets the eye. This is entirely different hardware, at first glance I see that Wi-Fi is handled by RTL8192E, and router itself is based on Qualcomm Atheros chipset. I don't see on which one though, because of photo resolution.
> > It looks like UART is right over USB socket on your board, but you will have to probe this yourself. And I thought that it only differs by a few unpopulated components...
> > I guess I'll edit the commit message to remove this reference, perhaps there is yet another router which has missing components from my board populated.
> 
> Ill try to get a more close-up shot of the board so that you may see wich chipsets are present on the board . but does that also mean that your openwrt project can't support the mf283V due to so much differences from the MF283+ ?

No, it only means, that you will have to do the reverse engineering an porting yourself, Qualcomm Atheros chipsets are very well supported in ath79 target. You might have issues with the modem though, it looks totally unknown to me and it's possible that you will need to add USB IDs to the kernel just like I did.

**Comment by Leo-PL** [comment]
@adschm one more question came to my mind. While I did the backport for 19.07 I noticed, that "+" sign is usable in device tree compatible string, but I'm not sure about the build system - should I keep the current `Device/zte_mf283plus` notation, or maybe change it to 'zte_mf283+`?

**Comment by KazuDante89** [comment]
> > > Wow. I'm quite surprised to see this. There's more to that than meets the eye. This is entirely different hardware, at first glance I see that Wi-Fi is handled by RTL8192E, and router itself is based on Qualcomm Atheros chipset. I don't see on which one though, because of photo resolution.
> > > It looks like UART is right over USB socket on your board, but you will have to probe this yourself. And I thought that it only differs by a few unpopulated components...
> > > I guess I'll edit the commit message to remove this reference, perhaps there is yet another router which has missing components from my board populated.
> > 
> > 
> > Ill try to get a more close-up shot of the board so that you may see wich chipsets are present on the board . but does that also mean that your openwrt project can't support the mf283V due to so much differences from the MF283+ ?
> 
> No, it only means, that you will have to do the reverse engineering an porting yourself, Qualcomm Atheros chipsets are very well supported in ath79 target. You might have issues with the modem though, it looks totally unknown to me and it's possible that you will need to add USB IDs to the kernel just like I did.

I have no knowledge in this sadly . so when I came across this thread it was the only hope I had left for this device . maybe if somehow I can gain full access to modify some configs to improve it .
l


> > > Wow. I'm quite surprised to see this. There's more to that than meets the eye. This is entirely different hardware, at first glance I see that Wi-Fi is handled by RTL8192E, and router itself is based on Qualcomm Atheros chipset. I don't see on which one though, because of photo resolution.
> > > It looks like UART is right over USB socket on your board, but you will have to probe this yourself. And I thought that it only differs by a few unpopulated components...
> > > I guess I'll edit the commit message to remove this reference, perhaps there is yet another router which has missing components from my board populated.
> > 
> > 
> > Ill try to get a more close-up shot of the board so that you may see wich chipsets are present on the board . but does that also mean that your openwrt project can't support the mf283V due to so much differences from the MF283+ ?
> 
> No, it only means, that you will have to do the reverse engineering an porting yourself, Qualcomm Atheros chipsets are very well supported in ath79 target. You might have issues with the modem though, it looks totally unknown to me and it's possible that you will need to add USB IDs to the kernel just like I did.

I have no knowledge in this at all sadly . only if this router was the mf283+ as well then maybe openwrt could of improved it if its not a hardware issue . maybe my beat chance is to find a way to gain full r/w access and edit some conf files if any in the firmware . but please refer me on amazon.com what I would need to probe it , when I endup getting the other modem then I can fully mess around with this board just in case something goes wrong and I end up bricking it

**Comment by Leo-PL** [comment]
Rebased against current master.
@KazuDante89 I used plain USB-UART adapter based on CP2102 or FT232RL chipset (probably fake one). Anything will work, just make sure it's using 3.3V I/O, and not +/-12V levels.

**Comment by adschm** [comment]
>  one more question came to my mind. While I did the backport for 19.07 I noticed, that "+" sign is usable in device tree compatible string, but I'm not sure about the build system - should I keep the current Device/zte_mf283plus notation, or maybe change it to 'zte_mf283+`?

I prefer _not_ using the '+' sign in the compatible.

**Comment by Leo-PL** [comment]
Since that was my initial feeling too, let's keep it that way.

**Comment by Leo-PL** [comment]
I just found out, that it might be possible to perform recovery, and so OpenWrt installation without disassembling the router, by pressing WPS + Reset together during power-on to trigger TFTP upgrade. Converting to draft for now until I can confirm.

**Comment by Leo-PL** [comment]
Installation through "TFTP recovery" triggered via buttons works indeed. I also found out that both PHONE1 and PHONE2 ports work and are tied together - Updated the commit message and marked as ready once again.

**Comment by Leo-PL** [comment]
1st patch for modem support landed in `master` and `net`, sent patch for qmi_wwan with Acked-by: from @bmork again.

**Comment by Leo-PL** [comment]
@adschm and 2nd patch landed too: https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net.git/commit/?id=88eee9b7b42e

**Comment by KazuDante89** [comment]
@Leo-PL  off-subject question , if you are using external/outdoor antenna's on your MF283+ , which antennas works for your device ?

**Comment by Leo-PL** [comment]
> 
> 
> @Leo-PL off-subject question , if you are using external/outdoor antenna's on your MF283+ , which antennas works for your device ?

I have pretty good coverage where I live, so internal LTE antennas work reasonably well, haven't tried "standard" external antennas yet, but they probably won't improve situation. Still, a WDR4300 + E3372s combo managed to get higher speeds - 30/18Mbps DL/UL on ZTE vs 45/20Mbps on the combo, standing side by side in the window. Adding a directional "cantenna" to both of them only worsened throughput due to lot of tall buildings in the area, so lot of reflections too.

Back on the topic, @adschm, could you please resolve your review comments? I suspect that "changes requested" might scare other reviewers off.

File: target/linux/ramips/image/rt305x.mk
Comment (adschm) [inline target/linux/ramips/image/rt305x.mk:1191]: You have no USB led?

File: target/linux/ramips/image/rt305x.mk
Comment (adschm) [inline target/linux/ramips/image/rt305x.mk:1192]: I found no other devices using kmod-usb-net-qmi-wwan or uqmi in DEVICE_PACKAGES. Do you really need it?

**Comment by adschm** [comment]
Since both patches are in linux-next, maybe replace the current patches with the upstream ones and move them to the 0xxx range of patches.

If it is fair to expect them to land in 5.12, add that to the patch name like done in generic (didn't check the number):
0123-v5.12-usb-serial-option-add-full-support-for-ZTE-P685M.patch

With that and the package questions resolved, this should be good for merge.

File: target/linux/ramips/image/rt305x.mk
Comment (adschm) [reply target/linux/ramips/image/rt305x.mk:1192]: My statement is apparantly wrong, looks like I did a mistake during searching earlier.

**Comment by adschm** [comment]
Patches have landed in kernel, I will update and merge this myself.
If you would have liked to keep usb-ledtrig for some reason, please open a new PR for changing that.

**Comment by Leo-PL** [comment]
@adschm Acked, kmod-ledtrig-usb wasn't needed. I was about to push the changes yesterday, and I would if I hadn't managed to cut myself off from the device on update, while being away from home.

**Comment by Leo-PL** [comment]
And thanks a lot for your support.
I'll happily test the build from your tree, to ensure everything works, of course after I return home and test what went wrong ;-)
Just pro forma, I can push my update now, too.

**Comment by adschm** [comment]
@Leo-PL This is already merged to master, as no functional changes have been made except the ledtrig removal.

**Comment by Leo-PL** [comment]
Great, thanks again!

**Comment by Leo-PL** [comment]
@adschm I found the problem - unfortunately snapshots are affected, too, for some reason image size stored inside sysupgrade image is too small: 
3: System Boot system code via Flash.
## Booting image at bc080000 ...
raspi_read: from:80000 len:40 
.   Image Name:   MIPS OpenWrt Linux-5.4.100
   Created:      2021-02-26  20:11:19 UTC
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    1626853 Bytes =  1.6 MB
   Load Address: 80000000
   Entry Point:  80000000
raspi_read: from:80040 len:18d2e5 
.........................   Verifying Checksum ... OK
   Uncompressing Kernel Image ... LZMA ERROR 1 - must RESET board to recover

This is 100% reproducible.
I wonder why this worked for me before removing kmod-led-trigger-usb and rebasing patches, and why images on my backport to 19.07 work too. 
I'll fix it and post a PR soon.

**Pull request closed without merging**
